### PR TITLE
Proposal for expansion of expression language

### PR DESCRIPTION
**Pull request by jaredoconnell**
## Changes introduced with this PR

This PR proposes three main changes to the expression language.

See the files changed for the proposal itself.

## Voting Period is ~~7~~ 21 days
Voting ends Friday ~~March 24th~~ April 7th.

---
By contributing to this repository, I agree to the [contribution guidelines](https://github.com/arcalot/.github/blob/main/CONTRIBUTING.md).

### CODE REVIEW

**Comment by jaredoconnell** [comment]
> I would like to add one more element: every expression language component should be built in such a way that the expression language can create a resulting schema without executing the expression itself if a schema for the referenced value is available. This will help with later adding static workflow analysis.

Especially since you're already implementing it, and it is required regardless as an implementation detail, I'll leave that out of the proposal. The proposed changes in this proposal change how the users create and interact with workflows.

**Review by sandrobonazzola — APPROVED** [review]

**Review by jdowni000 — APPROVED** [review]

**Comment by dustinblack** [comment]
I assume this is a breaking change, particularly:

> - Make subexpressions default within map access since it is unintuitive as-is, removing the old () syntax. Moving from `map[(sub-expression)]` to `map[sub-expression]`

The new functions and operators I suppose could add functionality in a non-breaking way.

IIUC this would cause a workflow built before the change to be incompatible with the engine after the change, and new workflows will not be backward-compatible. This will require some housekeeping across the workflows we already have published, and a good set of release notes. Documentation should be a concurrent requirement with the new engine release, when ready.

**Comment by jaredoconnell** [comment]
> I assume this is a breaking change, particularly:
> 
> > * Make subexpressions default within map access since it is unintuitive as-is, removing the old () syntax. Moving from `map[(sub-expression)]` to `map[sub-expression]`
> 
> The new functions and operators I suppose could add functionality in a non-breaking way.
> 
> IIUC this would cause a workflow built before the change to be incompatible with the engine after the change, and new workflows will not be backward-compatible. This will require some housekeeping across the workflows we already have published, and a good set of release notes. Documentation should be a concurrent requirement with the new engine release, when ready.

While the () removal is likely a breaking change, no one has used it as far as I am aware. It's an advanced use case that has not been taken advantage of.

The main documentation change is that this deviates from JSONPATH.

**Review by dustinblack — APPROVED** [review]

**Review by lmilbaum — APPROVED** [review]

File: art-decisions/proposals/2023-03-14-expression-expansion.md [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+- [Functions](https://github.com/arcalot/arcaflow-expressions/issues/1)
+- [Binary Comparison Operators](https://github.com/arcalot/arcaflow-expressions/issues/2)
+- [Mathematical Expressions](https://github.com/arcalot/arcaflow-expressions/issues/3)
+
+### Priorities 
+- The function calls and comparison operators will be the top priorities
+- Mathematical expressions will be lower priority, with most of the work based on the work for the comparison operators.
+
+### Approval
+By approving this proposal, you are agree to the addition of functions, binary comparison operators, and mathematical expressions to the language. Not necessarily the way it's implemented.
```
[the remark this answers is not in the document]
Comment (jaredoconnell) [reply art-decisions/proposals/2023-03-14-expression-expansion.md]: Thanks for pointing those out! Fixed.

**Review by tsebastiani — APPROVED** [review]

**Review by HubertStefanski — APPROVED** [review]

**Comment by dustinblack** [comment]
I'm taking the liberty to extend the voting date on this to April 7 since the process is still new for us and many of us were distracted by an internal event.

**Review by mfleader — APPROVED** [review]

**Pull request merged**